### PR TITLE
[wgsl] Describe switch statements

### LINKED ISSUE #630: [wgsl] propose switch-case permit multiple literal values per case
Right now the case statement allows matching only one literal value:

```
  const x_9 : u32 = w;
  switch (x_9) {
    default: {
      w = 10;
    }
    case 0: {
      w = 1;
    }
    case 1: {
      w = 2;
    }
    case 5: {
      fallthrough;
    }
    case 6: {
      w = 6;
    }
  }
```

In this example, both the 5 and 6 cases should do the same thing.    I propose that we be able to list multiple values per case statement, like this:

```
  const x_9 : u32 = w;
  switch (x_9) {
    default: {
      w = 10;
    }
    case 0: {
      w = 1;
    }
    case 1: {
      w = 2;
    }
    case 5, 6: {    // This is the changed line.
      w = 6;
    }
  }
```

This eliminates the need to write a bunch of boilerplate, namely `{ fallthrough; }`

It's also easier to read, and to preserve in a round-trip with SPIR-V and LLVM.  (In SPIR-V and LLVM, you can repeat the target block name for different case values.)

We would still retain the restriction that a case value may not appear in two different case statements for a particular switch construct.

### PR DESCRIPTION
**Pull request by dneto0**
Fixes #630

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -1039,15 +1039,42 @@ switch_stmt
...
+depending the evaluation of a selector expression of a scalar integer type.
+
+If the selector value equals a value in a case selector list, then control is transferred to
+the body of that case clause.
+If the selector value does not equal any of the case selector values, then control is
+transferred to the `default` clause.
+
+TODO(dneto): Is the default clause required? That would avoid undefined behaviour.
+
+A literal value must not appear more than once in the case selectors for a switch statement.
```
Comment (dj2) [inline wgsl/index.bs:1064]: Does the type of the literal value have to match the type of the switch conditional? I think yes, but it should probably be called out here.

File: wgsl/index.bs [hunk 2]
```diff
@@ -1039,15 +1039,42 @@ switch_stmt
...
+than one case selector list.
+
+Note: The value of the literal is what matters, not the spelling. For example `0` and `00` both
+denote the zero value.
+
+When control reaches the end of a case body, control normally transfers to the first statement
+after the switch statement.
+Alternately, executing a `fallthrough` statement transfers control to the body of the next case clause or
+default clause, whichever appears next in the switch body.
+If there is no such clause, then `fallthrough` transfers control to the first statement after the switch statement.
```
Comment (dj2) [inline wgsl/index.bs:1076]: This feels like it should be an error?

Comment (dneto0) [reply wgsl/index.bs:1076]: That could be a good choice.
I leaned slightly toward what I wrote to retain generality for an automatic generator.  But I don't have that use case particularly.

File: wgsl/index.bs [hunk 3]
```diff
@@ -1039,15 +1039,42 @@ switch_stmt
...
 
+A switch statement transfers control to one of a set of case clauses, or to a default clause,
+depending the evaluation of a selector expression of a scalar integer type.
+
+If the selector value equals a value in a case selector list, then control is transferred to
+the body of that case clause.
+If the selector value does not equal any of the case selector values, then control is
+transferred to the `default` clause.
+
+TODO(dneto): Is the default clause required? That would avoid undefined behaviour.
```
Comment (kvark) [inline wgsl/index.bs:1062]: Does bikeshed understand TODO clauses? I know it does for "Issue: Xxx", so maybe let's use that instead.

Comment (dneto0) [reply wgsl/index.bs]: Resolved the TODO anyway.

File: wgsl/index.bs (hunk 2 above)
Comment (dneto0) [reply wgsl/index.bs:1074]: I looked up what Swift says about fallthrough, and I read it as non-committal at best on this point.

https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html#ID140

File: wgsl/index.bs (hunk 1 above)
Comment (dneto0) [reply wgsl/index.bs:1062]: Added a sentence for this.

**Comment by dneto0** [comment]
I think the last remaining issue is from https://github.com/gpuweb/gpuweb/pull/713#discussion_r409165436: Is it valid to have a `fallthrough` statement in the last clause of a switch?  The current text says it transfers control to the first statement after the switch statement.

**Comment by grorg** [comment]
This was discussed at the [2020-04-28 meeting](https://docs.google.com/document/d/1zUMz2oU5ovhxK_qx4MB8wKFZbOmOdOXU-LPLkrLl9go/).

**Comment by dneto0** [comment]
Resolution from meeting: ban fallthrough in the last clause, as it likely indicates an error.
In future we can relax the rule.

**Comment by dj2** [comment]
Approved in the meeting, merging.

**Pull request merged**